mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -284,48 +294,69 @@ class CastConversion : public OpConversionPattern<ArithOp> {
                   ConversionPatternRewriter &rewriter) const override {
 
     Type opReturnType = this->getTypeConverter()->convertType(op.getType());
-    if (!isa_and_nonnull<IntegerType>(opReturnType)) {
-      return rewriter.notifyMatchFailure(op, "expected integer result type");
-    }
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
+            opReturnType))
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t result type");
 
     if (adaptor.getOperands().size() != 1) {
       return rewriter.notifyMatchFailure(
           op, "CastConversion only supports unary ops");
     }
 
     Type operandType = adaptor.getIn().getType();
-    if (!isa_and_nonnull<IntegerType>(operandType)) {
-      return rewriter.notifyMatchFailure(op, "expected integer operand type");
-    }
+    if (!isa_and_nonnull<IntegerType, emitc::SignedSizeTType, emitc::SizeTType>(
+            operandType))
+      return rewriter.notifyMatchFailure(
+          op, "expected integer or size_t/ssize_t operand type");
 
-    bool isTruncation = operandType.getIntOrFloatBitWidth() >
-                        opReturnType.getIntOrFloatBitWidth();
-    bool doUnsigned = needsUnsigned || isTruncation;
+    bool isTruncation =
```

Comment:
There might be truncation when going to size_t types even though we cannot find it our here. Why does that not matter?

Reply:
I'm afraid there's nothing we can do for these; we can't spot the case of going between signed index and signed integer, where destination is strictly shorter than source. Bugs will appear if the pre-C++20 behavior (which is implementation-defined) happens to be something else than truncation.

The safest thing we can do is to prevent `index_cast`... but that's probably not what we want.

Reply:
Then it's something we should look out for in our end2end testing

Reply:
But if we need to rely on that being truncation, do we need the isTruncation flag at all?

Reply:
After discussion -- the behavior of most compilers being to issue a truncation anyway, we can just proceed with this.

Relative to the isTruncation flag: it is true if we're sure it is a truncation (therefore using unsigned types is OK), but it being false doesn't prevent truncation from happening. I think we should keep it to avoid relying on implementation-defined behavior as much as possible.